### ISSUE TITLE
"Block-scoped variable used before its declaration" not detected inside closures

### ISSUE THREAD

**Issue by kourge** [microsoft/TypeScript#9273]
This is similar to #7477 and #8976, but occurs inside closures.

**TypeScript Version:** 

1.9.0-dev.20160620-1.0

**Code**

``` ts
const y = (() => x)();
const x = 3;
```

**Expected behavior:**

The compilation error "Block-scoped variable `x` used before its declaration".

**Actual behavior:** 

The code is compiled to:

``` ts
var y = (function () { return x; })();
var x = 3;
```

This sets `y` to `undefined`.

**Additional variants:**

This also occurs when the access to the binding `x` is in default argument value position:

``` ts
const y = ((i: number = x) => i)();
const x = 3;
```

**Comparison against spec behavior:**
All of the following result in a `ReferenceError` being thrown on Node 5.1:

``` js
(function() {
  'use strict';
  var y = (() => x)();
  const x = 3;
  return y;
})();

(function() {
  'use strict';
  const y = (() => x)();
  const x = 3;
  return y;
})();
```

**Comment by RyanCavanaugh** [comment]
Consider the dual of this, #9757, where someone *didn't* want an error even when the variable was initialized with a reference to itself.

We could special-case IIFEs to be considered inlined for the purposes of flow analysis, but in general we can't tell whether a given function expression executes "now" or "later", so that'd just be a patch, but it might be one worth making.

**Comment by kourge** [comment]
IIFEs are a very common pattern, and additionally they are distinguishable on a syntactic level. It would be a great boon for flow analysis to consider them in the special way, however limited this exception may be. But like you said, not being able to tell when a function expression gets invoked is a great limitation, and might even step into the realm of escape analysis.

**Comment by yortus** [comment]
> IIFEs are a very common pattern [....] It would be a great boon for flow analysis to consider them in the special way

#8849 already implements this, so it's surprising that it doesn't catch use before declaration. Consider the same example but with a union type for `x`, and another identical IIFE just after the assignment:

```ts
const y = (() => x)(); // x is string|number, due to #8849 TS *knows* it hasn't been assigned
                       // when the IIFE runs (although unassigned consts can't exist!)

const x: string|number = 3;

const z = (() => x)(); // x is number, due to #8849 TS *knows* it must
                       // be a number when the IIFE runs
```


This shows that the compiler *has* treated the IIFEs as part of the surrounding control flow for the purposes of narrowing, but doesn't appear to recognise use before declaration of `x`.

**Comment by RyanCavanaugh** [comment]
This is fixed now
